File: Graph/d.SatisfatibilityOfEquations.cpp
class Solution {
public:
    int * parent;
    int * ranker;
    int find(int x) {
        if(parent[x] == x){
            return x;
        }else {
            parent[x] = find(parent[x]);
            return parent[x];
        }
    }
    void Union(int xl , int yl) {
        if(ranker[xl] < ranker[yl]){
            parent[xl] = yl;
        }else if(ranker[yl] < ranker[xl]){
            parent[yl] = xl;
        }else {
            parent[xl] = yl;
            ranker[yl]++;
        }
    }
    bool equationsPossible(vector<string>& equations) {
        parent = new int[26];
        ranker = new int[26];
        for(int i = 0 ; i < 26 ; i++) {
            parent[i] = i;
            ranker[i] = 0;
        }
        for(string equation : equations) {
            if(equation[1] == '='){
                int op1 = equation[0] - 'a';
                int op2 = equation[3] - 'a';
                int l1 = find(op1);
                int l2 = find(op2);
                Union(l1 , l2);
            }
        }
        
        for(string equation : equations) {
            if(equation[1] == '!'){
                int op1 = equation[0] - 'a';
                int op2 = equation[3] - 'a';
                int l1 = find(op1);
                int l2 = find(op2);
                if(l1 == l2){
                    return false;
                }
            }
        }
        return true;
    }
};

File: Graph/d.SatisfatibilityOfEquations.java
class Solution {
    int[]parent;
    int[]rank;
    public int find(int x){
        if(parent[x] == x){
            return x;
        }else{
            parent[x] = find(parent[x]);
            return parent[x];
        }
    }
    public void union(int xl , int yl){
        if(rank[xl] < rank[yl]){
            parent[xl] = yl;
        }else if(rank[yl] < rank[xl]){
            parent[yl] = xl;
        }else {
            parent[xl] = yl;
            rank[yl]++;
        }
    }
    public boolean equationsPossible(String[] equations) {
        parent = new int[26];
        rank = new int[26];
        for(int i = 0 ; i < 26 ; i++){
            parent[i] = i;
            rank[i] = 0;
        }
        for(String equation : equations){
            if(equation.charAt(1) == '=' ){
                int op1 = equation.charAt(0) - 'a';
                int op2 = equation.charAt(3) - 'a';
                int l1 = find(op1);
                int l2 = find(op2);
                if(l1 != l2){
                    union(l1 , l2);
                }
            }
        }
        for(String equation : equations) {
            if(equation.charAt(1) == '!'){
                int op1 = equation.charAt(0) - 'a';
                int op2 = equation.charAt(3) - 'a';
                int l1 = find(op1);
                int l2 = find(op2);
                if(l1 == l2){
                    return false;
                } else {
                    continue;
                }
            }
        }
        return true;
    }
}